### PR TITLE
feat(support-x-interservice-auth-header): Add support for x-interservice-auth header

### PR DESCRIPTION
**Pull request by papey**
# Neurow

## Context 🧐

Authorization header for interservice communication should use "x-interservice-authorization" header key rather than the classic "authorization" header.

## Changes ⛏️

- Added a double check on both header keys, new one as precedence
- Added a test to ensure we prefer the new key over the old one

### CODE REVIEW

File: neurow/lib/neurow/jwt_auth_plug.ex [hunk 1]
```diff
@@ -124,13 +124,20 @@ defmodule Neurow.JwtAuthPlug do
     end)
   end
 
-  defp jwt_token_from_request(conn) do
-    case conn |> get_req_header("authorization") do
-      ["Bearer " <> jwt_token] ->
-        {:ok, jwt_token}
+  @auth_header_keys ["x-interservice-authorization", "authorization"]
```
Comment (achouippe) [inline neurow/lib/neurow/jwt_auth_plug.ex:127]: For the public API, we want to continue using the regular `authorization` HTTP header. `x-interservice-authorization` should only be used for backend to backend communications.

I think we should pass the name of the authorization header as a configuration parameter when jwt_auth_plug is added to plug pipeline.

**Review by achouippe — CHANGES_REQUESTED** [review]

File: neurow/lib/neurow/jwt_auth_plug.ex (hunk 1 above)
Comment (papey) [reply neurow/lib/neurow/jwt_auth_plug.ex:127]: oops yes, get it!

Comment (papey) [reply neurow/lib/neurow/jwt_auth_plug.ex]: fixed ✅

File: neurow/test/neurow/internal_api/endpoint_test.exs [hunk 2]
```diff
@@ -31,6 +31,19 @@ defmodule Neurow.InternalApi.EndpointTest do
     assert call.status == 404
   end
 
+  test "other routes requires a JWT sent in x-interservice-authorization header" do
```
Comment (papey) [inline neurow/test/neurow/internal_api/endpoint_test.exs:34]: @achouippe this ensure we can process both headers, once migration is done, we will fallback to only one key and remove the need for the array if thats ok for you.

**Comment by achouippe** [comment]
> @achouippe another question, I think header should start with a capital letter, eg: `Authorization` is prefer over `authorization` (HTTP/1.1 specification)

If you use http headers with capital letter, `Plug` is not happy. It automatically translates header names to lowercase.

**Comment by papey** [comment]
> > @achouippe another question, I think header should start with a capital letter, eg: `Authorization` is prefer over `authorization` (HTTP/1.1 specification)
> 
> If you use http headers with capital letter, `Plug` is not happy. It automatically translates header names to lowercase.

ok then 👍

File: neurow/lib/neurow/internal_api/endpoint.ex [hunk 3]
```diff
@@ -9,6 +9,7 @@ defmodule Neurow.InternalApi.Endpoint do
   plug(MetricsPlugExporter)
 
   plug(Neurow.JwtAuthPlug,
+    auth_header_keys: ["x-interservice-authorization", "authorization"],
```
Comment (papey) [inline neurow/lib/neurow/internal_api/endpoint.ex:12]: 📓 : the fallback is here

Comment (achouippe) [reply neurow/lib/neurow/internal_api/endpoint.ex:12]: Nit: Name this property `credential_headers`

File: neurow/lib/neurow/jwt_auth_plug.ex [hunk 4]
```diff
@@ -124,13 +125,16 @@ defmodule Neurow.JwtAuthPlug do
...
-    case conn |> get_req_header("authorization") do
-      ["Bearer " <> jwt_token] ->
-        {:ok, jwt_token}
-
-      _ ->
-        {:error, :invalid_authorization_header, "Invalid authorization header"}
+  defp jwt_token_from_request(conn, options) do
+    Enum.find_value(options.auth_header_keys, fn header ->
+      case get_req_header(conn, header) do
+        ["Bearer " <> jwt_token] -> jwt_token
```
Comment (achouippe) [inline neurow/lib/neurow/jwt_auth_plug.ex:131]: Question: Is this also a Bearer with `x-interservice-authorization` ?

File: neurow/test/jwt_helper.exs [hunk 5]
```diff
@@ -16,10 +16,14 @@ defmodule JwtHelper do
     conn |> put_req_header("authorization", "Bearer #{jwt_token}")
   end
 
-  def put_jwt_token_in_req_header_internal_api(conn, issuer \\ "test_issuer1") do
+  def put_jwt_token_in_req_header_internal_api(
+        conn,
+        issuer \\ "test_issuer1",
+        header_key \\ "authorization"
```
Comment (achouippe) [inline neurow/test/jwt_helper.exs:22]: Nit: `header_name`

**Review by achouippe — COMMENTED** [review]
Some nits and 1 question.

**Review by achouippe — APPROVED** [review]

File: neurow/lib/neurow/jwt_auth_plug.ex (hunk 4 above)
Comment (papey) [reply neurow/lib/neurow/jwt_auth_plug.ex:131]: Yes, same format, just different name

File: neurow/lib/neurow/internal_api/endpoint.ex (hunk 3 above)
Comment (papey) [reply neurow/lib/neurow/internal_api/endpoint.ex]: Done ✅

File: neurow/test/jwt_helper.exs (hunk 5 above)
Comment (papey) [reply neurow/test/jwt_helper.exs]: Done ✅

**Pull request merged**